substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrTypeRepository.java

```diff
@@ -66,7 +66,7 @@ public int write(JfrChunkWriter writer) {
     private TypeInfo collectTypeInfo() {
         TypeInfo typeInfo = new TypeInfo();
         for (Class<?> clazz : Heap.getHeap().getLoadedClasses()) {
-            if (JfrTraceId.isUsedPreviousEpoch(clazz)) {
+            if (JfrTraceId.isUsedCurrentEpoch(clazz)) {
```

Comment:
During the writing of the stack trace repository((closeFile -> writeCheckPoints -> writeRepositories -> write), I'm iterating through all methods in that stack trace. `Jfr.Method` type contains among other information, information about Class. So I'm calling `typeRepo.getClassId(clazz)` (https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrStackTraceRepository.java#L175-L175). The problem here is that if I call `getClassId` from the write repository method, it will be registered as used in the current epoch and not in the previous, which further means that non of JFR reports will contain information about method's class.

If you have any ideas how we can solve this easily, we can try.

Reply:
To get the class id when writing out the constant pools, you should use `JfrTraceId.getTraceId(Class<?> clazz)`. This is what is used when writing out the constant pool entries for the classes themselves (https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrTypeRepository.java#L118)

Reply:
It is not correct to alter `collectTypeInfo()` this way; it will then output an unrelated set of constant pool entries that do not match the events in the chunk being closed.

Reply:
I guess following, the `JfrStackTraceRepository` needs to be "epoch-aware" like the `JfrSymbolRepository` where there are two tables, one for each epoch. When an entry for the repository is marked as in use, it should also mark any child entries (like methods and their classes) as in use at the same time.

Reply:
Implementation is now slightly different. `JfrStackTraceRepository` is marking all classes from stack trace as [visited](https://github.com/oracle/graal/pull/3720/files#diff-db688afba98414edb0d2cc0d0cecc6a3d6900766f46cf5b05418b8d68ecddcc5R80-R85) and then, during the writing `JfrTraceId.getTraceId(Class<?> clazz)` will [return class id](https://github.com/oracle/graal/pull/3720/files#diff-15a00b6f6d7993eea19b0d110f130835c03d5fe4ac6ad82bb6147f97f81a3d08R79), for class that is already loaded in the previous epoch. So there is no need for changes in `collectTypeInfo`.